### PR TITLE
Azure: Limit polling network metadata on connection errors

### PR DESCRIPTION
**Pull request by aswinrajamannar**
## Proposed Commit Message

```
Azure: Limit polling network metadata on connection errors
```
## Description
This is a follow up to #878 where we currently limit polling on network metadata to 10 retries if the calls fail due to timeout errors. After enabling the feature in Azure, we noticed that if secondary nic is attached first, the calls could also fail with network unreachable ConnectionError. In these cases as well, we shouldn't poll indefinitely, and instead move on to listen for more nic attach events.

## Test Steps
- Deploy a multi-nic VM in Azure that goes through the nic hot-attach code path. In a few cases where the secondary nic is attached first, the VM deployment might fail if this fix is missing.

## Checklist:
 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/hacking.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

### CODE REVIEW

File: cloudinit/sources/DataSourceAzure.py [hunk 1]
```diff
@@ -1011,9 +1014,12 @@ def network_metadata_exc_cb(msg, exc):
                                             (msg, exc.cause, exc.code),
                                             logger_func=LOG.error)
 
-            if exc.cause and isinstance(exc.cause, requests.Timeout):
-                metadata_timeout_count = metadata_timeout_count + 1
-                return (metadata_timeout_count <= 10)
+            # Retry up to a certain limit for both timeout and network
+            # unreachable errors.
+            if exc.cause and (isinstance(exc.cause, requests.Timeout) or
+                              isinstance(exc.cause, requests.ConnectionError)):
```
Comment (TheRealFalcon) [inline cloudinit/sources/DataSourceAzure.py:1020]:
```suggestion
            if exc.cause and isinstance(
                exc.cause, (requests.Timeout, requests.ConnectionError)
            ):
```

File: tests/unittests/test_datasource/test_azure.py [hunk 2]
```diff
@@ -2854,10 +2855,17 @@ def network_metadata_ret(ifname, retries, type, exc_cb, infinite):
...
+                    # We are expected to retry for a certain period for both
+                    # timeout errors and network unreachable errors.
+                    if _ < 5:
+                        cause = requests.Timeout('Fake connection timeout')
+                        error = url_helper.UrlError(cause=cause)
+                        eth1Retries.append(exc_cb("Connection timeout", error))
+                    else:
+                        cause = requests.ConnectionError('Network Unreachable')
+                        error = url_helper.UrlError(cause=cause)
+                        eth1Retries.append(exc_cb("Connection timeout", error))
```
Comment (TheRealFalcon) [inline tests/unittests/test_datasource/test_azure.py:2868]:
```suggestion
                    if _ < 5:
                        cause = requests.Timeout('Fake connection timeout')
                    else:
                        cause = requests.ConnectionError('Network Unreachable')
                    error = url_helper.UrlError(cause=cause)
                    eth1Retries.append(exc_cb("Connection timeout", error))
```

**Review by TheRealFalcon — COMMENTED** [review]
Looks good to me! I left two nit suggestions inline.

**Review by TheRealFalcon — APPROVED** [review]

**Comment by aswinrajamannar** [comment]
> Looks good to me! I left two nit suggestions inline.

Thanks. Committed the suggestions.

**Pull request merged**